File: sol/solution/0101-0200/0106/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {Integer[]} inorder
# @param {Integer[]} postorder
# @return {TreeNode}
def build_tree(inorder, postorder)
  @in_map = {}
  inorder.each_with_index { |val, i| @in_map[val] = i }
  @post_index = postorder.length - 1
  build_tree_helper(inorder, postorder, 0, inorder.length - 1)
end
def build_tree_helper(inorder, postorder, in_start, in_end)
  return nil if in_start > in_end
  root = TreeNode.new(postorder[@post_index])
  @post_index -= 1
  in_index = @in_map[root.val]
  root.right = build_tree_helper(inorder, postorder, in_index + 1, in_end)
  root.left = build_tree_helper(inorder, postorder, in_start, in_index - 1)
  root
end

File: sol/solution/0101-0200/0106/Solution.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).
-spec build_tree(Inorder :: [integer()], Postorder :: [integer()]) -> #tree_node{} | null.
build_tree(Inorder, Postorder) ->
    InMap = maps:from_list(lists:zip(Inorder, lists:seq(0, length(Inorder) - 1))),
    PostIndex = length(Postorder) - 1,
    build_tree_helper(Inorder, Postorder, 0, length(Inorder) - 1, PostIndex, InMap).
build_tree_helper(_Inorder, _Postorder, InStart, InEnd, _PostIndex, _InMap) when InStart > InEnd ->
    null;
build_tree_helper(Inorder, Postorder, InStart, InEnd, PostIndex, InMap) ->
    RootVal = lists:nth(PostIndex + 1, Postorder),
    Root = #tree_node{val = RootVal},
    InIndex = maps:get(RootVal, InMap),
    RightSize = InEnd - InIndex,
    Right = build_tree_helper(Inorder, Postorder, InIndex + 1, InEnd, PostIndex - 1, InMap),
    Left = build_tree_helper(Inorder, Postorder, InStart, InIndex - 1, PostIndex - 1 - RightSize, InMap),
    Root#tree_node{left = Left, right = Right}.